marshmallow_recipe/generics.py

```diff
@@ -0,0 +1,120 @@
+else:
+    DataclassInstance: TypeAlias = type
+
+TypeLike: TypeAlias = type | TypeVar | types.UnionType | types.GenericAlias | _GenericAlias
+FieldsTypeMap: TypeAlias = dict[str, TypeLike]
+TypeVarMap: TypeAlias = dict[TypeVar, TypeLike]
+FieldsClassMap: TypeAlias = dict[str, TypeLike]
+ClassTypeVarMap: TypeAlias = dict[TypeLike, TypeVarMap]
+FieldsTypeVarMap: TypeAlias = dict[str, TypeVarMap]
+
+
+def get_fields_type_map(cls: type) -> FieldsTypeMap:
+    origin: type = get_origin(cls) or cls
+    if not dataclasses.is_dataclass(origin):
+        raise ValueError(f"{origin} is not a dataclass")
+
+    class_type_var_map = get_class_type_var_map(cls)
+    fields_class_map = get_fields_class_map(origin)
+    return {
+        f.name: build_subscripted_type(f.type, class_type_var_map.get(fields_class_map[f.name], {}))
+        for f in dataclasses.fields(origin)
+    }
+
+
+def get_fields_class_map(cls: type[DataclassInstance]) -> FieldsClassMap:
+    names: dict[str, dataclasses.Field] = {}
+    result: FieldsClassMap = {}
+
+    mro = cls.__mro__
+    for cls in (*mro[-1:0:-1], cls):
+        if not dataclasses.is_dataclass(cls):
+            continue
+        for field in dataclasses.fields(cls):
+            if names.get(field.name) != field:
+                names[field.name] = field
+                result[field.name] = cls
+
+    return result
+
+
+def build_subscripted_type(t: TypeLike, type_var_map: TypeVarMap) -> TypeLike:
+    if isinstance(t, TypeVar):
+        return build_subscripted_type(type_var_map[t], type_var_map)
+
+    origin = get_origin(t)
+    if origin is Union or origin is types.UnionType:
+        return Union[*(build_subscripted_type(x, type_var_map) for x in get_args(t))]
+
+    if origin is Annotated:
+        t, *annotations = get_args(t)
+        return Annotated[build_subscripted_type(t, type_var_map), *annotations]
+
+    if origin and isinstance(t, types.GenericAlias):
+        return types.GenericAlias(origin, tuple(build_subscripted_type(x, type_var_map) for x in get_args(t)))
+
+    if origin and isinstance(t, _GenericAlias):
+        return _GenericAlias(origin, tuple(build_subscripted_type(x, type_var_map) for x in get_args(t)))
+
+    return _subscript_with_any(t)
+
+
+def get_class_type_var_map(t: TypeLike) -> ClassTypeVarMap:
+    class_type_var_map: ClassTypeVarMap = {}
+    _build_class_type_var_map(t, class_type_var_map)
+    return class_type_var_map
+
+
+def _build_class_type_var_map(t: TypeLike, class_type_var_map: ClassTypeVarMap) -> None:
+    if _get_params(t):
+        raise ValueError(f"Expected subscripted generic, but got unsubscripted {t}")
+
+    type_var_map: TypeVarMap = {}
+    origin = get_origin(t) or t
+    params = _get_params(origin)
+    args = get_args(t)
+    if params or args:
+        if not params or not args or len(params) != len(args):
+            raise ValueError(f"Unexpected generic {t}")
+        class_type_var_map[origin] = type_var_map
```

Comment:
we override `type_var_map` for `origin`value and don't check whether it is already there (it could be there for cases of multiple inheritance, and in that case we receive wrong result as it is demonstrated in tests above). I would suggest to raise an exception in case if `origin` is already in `class_type_var_map`

Reply:
I have made some improvements https://github.com/anna-money/marshmallow-recipe/pull/163/commits/461e0d9d085793577f9e58257a7c79ff01da2822